subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy

```diff
@@ -187,67 +183,247 @@ class JavaToolchainBuildOperationsIntegrationTest extends AbstractIntegrationSpe
+        JvmInstallationMetadata jdkMetadata = AvailableJavaHomes.getJvmInstallationMetadata(AvailableJavaHomes.differentJdk)
+
+        buildFile << """
+            compileJava {
+                options.forkOptions.javaHome = file("${jdkMetadata.javaHome}")
+            }
+        """
+
+        file("src/main/java/Foo.java") << """
+            public class Foo { }
+        """
+
+        def task = ":compileJava"
+
+        when:
+        runWithInstallation(jdkMetadata, task)
+        def events = eventsFor(task)
+        then:
+        executedAndNotSkipped(task)
+        assertToolchainUsages(events, jdkMetadata, "JavaCompiler")
+
+        when:
+        runWithInstallation(jdkMetadata, task)
+        events = eventsFor(task)
+        then:
+        skipped(task)
+        assertToolchainUsages(events, jdkMetadata, "JavaCompiler")
+    }
+
+    @ToBeImplemented("finding a used toolchain by the path of the launcher executable")
+    def "emits toolchain usages for test that configures executable path"() {
+        JvmInstallationMetadata jdkMetadata = AvailableJavaHomes.getJvmInstallationMetadata(AvailableJavaHomes.differentVersion)
+
+        buildFile << """
+            compileTestJava {
+                javaCompiler = javaToolchains.compilerFor {
+                    languageVersion = JavaLanguageVersion.of(${jdkMetadata.languageVersion.majorVersion})
+                }
+            }
+
+            def javaExecutable = javaToolchains.launcherFor {
+                languageVersion = JavaLanguageVersion.of(${jdkMetadata.languageVersion.majorVersion})
+            }.get().executablePath
+
+            test {
+                executable = javaExecutable
+            }
+        """
+
+        file("src/test/java/FooTest.java") << """
+            public class FooTest {
+                @org.junit.Test
+                public void test() {}
+            }
+        """
+
+        def task = ":test"
+
+        when:
+        runWithInstallation(jdkMetadata, task)
+//        def events = eventsFor(task)
+        then:
+        executedAndNotSkipped(task)
+        // TODO: uncomment this when fixed
+//        assertToolchainUsages(events, jdkMetadata, "JavaLauncher")
+
+        when:
+        runWithInstallation(jdkMetadata, task)
+//        events = eventsFor(task)
+        then:
+        skipped(task)
+        // TODO: uncomment this when fixed
+//        assertToolchainUsages(events, jdkMetadata, "JavaLauncher")
+    }
+
+    def "emits toolchain usages when configuring toolchains for Kotlin plugin for '#task' task"() {
+        JvmInstallationMetadata jdkMetadata = AvailableJavaHomes.getJvmInstallationMetadata(AvailableJavaHomes.differentVersion)
+
+        // override setup
+        buildFile.text = """
+            buildscript {
+                ${mavenCentralRepository()}
+                dependencies { classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:1.7.10" }
```

Comment:
I am not sure it is the best solution to keep Kotlin version hardcoded here.
However, I wasn't able to quickly figure out if we have some variable that hold something like latest supported Kotlin plugin version or similar.

Reply:
We have a class `KotlinGradlePluginVersions` which you can use in a where clause to test multiple Kotlin versions.